lib/ClangImporter/SwiftDeclSynthesizer.cpp

```diff
@@ -2534,128 +2534,167 @@ SwiftDeclSynthesizer::makeDefaultArgument(const clang::ParmVarDecl *param,
+          clangCtx, synthesizedCxxMethodDecl, cxxRecordDeclLoc,
+          cxxRecordDeclLoc, origParam->getIdentifier(), origParam->getType(),
+          clangCtx.getTrivialTypeSourceInfo(origParam->getType()),
+          clang::SC_None, nullptr);
+      synthesizedParams.push_back(param);
+    }
+    synthesizedCxxMethodDecl->setParams(synthesizedParams);
+
+    if (!hasImmortalAttrs(cxxRecordDecl)) {
+      clang::SwiftAttrAttr *returnsRetainedAttrForSynthesizedCxxMethodDecl =
+          clang::SwiftAttrAttr::Create(clangCtx, "returns_retained");
+      synthesizedCxxMethodDecl->addAttr(
+          returnsRetainedAttrForSynthesizedCxxMethodDecl);
+    }
+
+    std::string swiftInitStr = "init(";
+    for (unsigned i = 0; i < ctorParamCount; ++i) {
+      swiftInitStr += "_:";
+    }
+    swiftInitStr += ")";
+    clang::SwiftNameAttr *swiftNameAttr =
+        clang::SwiftNameAttr::Create(clangCtx, swiftInitStr);
+    synthesizedCxxMethodDecl->addAttr(swiftNameAttr);
+
+    llvm::SmallVector<clang::Expr *, 8> ctorArgs;
+    for (auto *param : synthesizedParams) {
+      ctorArgs.push_back(clang::DeclRefExpr::Create(
+          clangCtx, clang::NestedNameSpecifierLoc(), clang::SourceLocation(),
+          param,
+          /*RefersToEnclosingVariableOrCapture=*/false, clang::SourceLocation(),
+          param->getType(), clang::VK_LValue));
+    }
+    llvm::SmallVector<clang::Expr *, 8> ctorArgsToAdd;
+    if (clangSema.CompleteConstructorCall(selectedCtorDecl, cxxRecordTy,
+                                          ctorArgs, cxxRecordDeclLoc,
+                                          ctorArgsToAdd))
+      continue;
+
+    clang::ExprResult synthesizedConstructExprResult =
+        clangSema.BuildCXXConstructExpr(
+            cxxRecordDeclLoc, cxxRecordTy, selectedCtorDecl,
+            /*Elidable=*/false, ctorArgsToAdd,
+            /*HadMultipleCandidates=*/false,
+            /*IsListInitialization=*/false,
+            /*IsStdInitListInitialization=*/false,
+            /*RequiresZeroInit=*/false, clang::CXXConstructionKind::Complete,
+            clang::SourceRange(cxxRecordDeclLoc, cxxRecordDeclLoc));
+    assert(!synthesizedConstructExprResult.isInvalid() &&
+           "Unable to synthesize constructor expression for c++ foreign "
+           "reference type");
+    clang::Expr *synthesizedConstructExpr =
+        synthesizedConstructExprResult.get();
+
+    clang::ExprResult synthesizedNewExprResult = clangSema.BuildCXXNew(
+        clang::SourceRange(), /*UseGlobal=*/false, clang::SourceLocation(), {},
+        clang::SourceLocation(), clang::SourceRange(), cxxRecordTy,
+        clangCtx.getTrivialTypeSourceInfo(cxxRecordTy), std::nullopt,
+        clang::SourceRange(cxxRecordDeclLoc, cxxRecordDeclLoc),
+        synthesizedConstructExpr);
+    assert(
+        !synthesizedNewExprResult.isInvalid() &&
+        "Unable to synthesize `new` expression for c++ foreign reference type");
+    clang::CXXNewExpr *synthesizedNewExpr =
```

Comment:
Nit: you can use auto here. The type is already there in the cast expression, you don't need to repeat it twice in the same statement.